File: src/domain/character.ts
import { createToyBaseObject, type BaseObject, type IsoDateTime, type StableId } from "./common";

export interface CharacterStableProfile extends BaseObject {
  type: "character-stable-profile";
  workId: StableId;
  displayName: string;
  role: string;
  coreDrive: string;
  coreDriveEvidenceRefIds: StableId[];
  fearOrWound: string;
  boundaries: string[];
  boundaryEvidenceRefIds: StableId[];
  competencies: string[];
  values: string[];
  voiceNotes: string[];
  voiceEvidenceRefIds: StableId[];
  actionFingerprints: string[];
}

export interface CharacterDynamicState extends BaseObject {
  type: "character-dynamic-state";
  characterId: StableId;
  scopeRef: StableId;
  timepoint: string;
  emotionalState: string;
  knowledgeState: string[];
  resources: string[];
  constraints: string[];
  evidenceRefIds: StableId[];
}

export interface CharacterArc extends BaseObject {
  type: "character-arc";
  characterId: StableId;
  arcGoal: string;
  arcGoalEvidenceRefIds: StableId[];
  startingBelief: string;
  pressurePattern: string;
  turningPoints: StableId[];
  currentStage: string;
}

export interface CharacterTimelineEntry {
  timepoint: string;
  eventRef: StableId;
  stateDelta: string;
  evidenceRef?: StableId;
  changeSetRef?: StableId;
}

export interface CharacterTimeline extends BaseObject {
  type: "character-timeline";
  characterId: StableId;
  entries: CharacterTimelineEntry[];
  lastVerifiedAt: IsoDateTime;
}

export type RelationshipTemperatureLevel = "cold" | "cool" | "neutral" | "warm" | "hot";

export interface RelationshipEdge extends BaseObject {
  type: "relationship-edge";
  workId: StableId;
  participantIds: StableId[];
  relationshipKind: string;
  currentTemperature: RelationshipTemperatureLevel;
  trust: number;
  tension: number;
  publicStatus: string;
  privateStatus: string;
  evidenceRefIds: StableId[];
}

export interface RelationshipTemperature extends BaseObject {
  type: "relationship-temperature";
  relationshipEdgeId: StableId;
  scopeRef: StableId;
  affinity: number;
  conflict: number;
  dependency: number;
  recentTriggers: string[];
}

export interface FocusRelationshipView extends BaseObject {
  type: "focus-relationship-view";
  relationshipEdgeId: StableId;
  taskType: string;
  includedStateRefs: StableId[];
  excludedStateRefs: StableId[];
  activationReason: string;
}

export interface InteractionPattern extends BaseObject {
  type: "interaction-pattern";
  participantRolePattern: string;
  trigger: string;
  typicalResponse: string;
  variationRules: string[];
  riskNotes: string[];
}

export function createToyCharacterProfile(timestamp: IsoDateTime): CharacterStableProfile {
  return {
    ...createToyBaseObject(
      "character-stable-profile",
      "character-practical-learner",
      "Practical Learner",
      timestamp,
      "active",
    ),
    workId: "work-public-toy",
    displayName: "Practical Learner",
    role: "protagonist",
    coreDrive: "Preserve the clue with the broadest future use.",
    coreDriveEvidenceRefIds: ["evidence-toy-pressure"],
    fearOrWound: "Worries that early certainty hides a later cost.",
    boundaries: ["Will not spend another person's resource without consent."],
    boundaryEvidenceRefIds: ["evidence-toy-pressure"],
    competencies: ["Careful observation", "Tool use"],
    values: ["Consent", "Practical truth"],
    voiceNotes: ["Short concrete sentences under pressure."],
    voiceEvidenceRefIds: ["evidence-toy-pressure"],
    actionFingerprints: ["Checks cost before acting."],
  };
}

export function createToyRelationshipEdge(timestamp: IsoDateTime): RelationshipEdge {
  return {
    ...createToyBaseObject(
      "relationship-edge",
      "relationship-mentor-learner",
      "Mentor and Learner",
      timestamp,
      "active",
    ),
    workId: "work-public-toy",
    participantIds: ["character-practical-learner", "character-careful-mentor"],
    relationshipKind: "mentor",
    currentTemperature: "warm",
    trust: 7,
    tension: 3,
    publicStatus: "Cooperative partners.",
    privateStatus: "The learner withholds uncertainty.",
    evidenceRefIds: ["evidence-toy-pressure"],
  };
}

File: crates/core/src/model/character.rs
use serde::{de::Error as DeserializeError, Deserialize, Deserializer, Serialize};
use serde_json::Number;

use super::common::{BaseObject, IsoDateTime, StableId};

pub type Score = Number;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterStableProfile {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_id: StableId,
    pub display_name: String,
    pub role: String,
    pub core_drive: String,
    pub core_drive_evidence_ref_ids: Vec<StableId>,
    pub fear_or_wound: String,
    pub boundaries: Vec<String>,
    pub boundary_evidence_ref_ids: Vec<StableId>,
    pub competencies: Vec<String>,
    pub values: Vec<String>,
    pub voice_notes: Vec<String>,
    pub voice_evidence_ref_ids: Vec<StableId>,
    pub action_fingerprints: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterDynamicState {
    #[serde(flatten)]
    pub base: BaseObject,
    pub character_id: StableId,
    pub scope_ref: StableId,
    pub timepoint: String,
    pub emotional_state: String,
    pub knowledge_state: Vec<String>,
    pub resources: Vec<String>,
    pub constraints: Vec<String>,
    pub evidence_ref_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterArc {
    #[serde(flatten)]
    pub base: BaseObject,
    pub character_id: StableId,
    pub arc_goal: String,
    pub arc_goal_evidence_ref_ids: Vec<StableId>,
    pub starting_belief: String,
    pub pressure_pattern: String,
    pub turning_points: Vec<StableId>,
    pub current_stage: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterTimelineEntry {
    pub timepoint: String,
    pub event_ref: StableId,
    pub state_delta: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub evidence_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub change_set_ref: Option<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterTimeline {
    #[serde(flatten)]
    pub base: BaseObject,
    pub character_id: StableId,
    pub entries: Vec<CharacterTimelineEntry>,
    pub last_verified_at: IsoDateTime,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationshipEdge {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_id: StableId,
    pub participant_ids: Vec<StableId>,
    pub relationship_kind: String,
    pub current_temperature: RelationshipTemperatureLevel,
    #[serde(deserialize_with = "deserialize_score")]
    pub trust: Score,
    #[serde(deserialize_with = "deserialize_score")]
    pub tension: Score,
    pub public_status: String,
    pub private_status: String,
    pub evidence_ref_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RelationshipTemperatureLevel {
    Cold,
    Cool,
    Neutral,
    Warm,
    Hot,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationshipTemperature {
    #[serde(flatten)]
    pub base: BaseObject,
    pub relationship_edge_id: StableId,
    pub scope_ref: StableId,
    #[serde(deserialize_with = "deserialize_score")]
    pub affinity: Score,
    #[serde(deserialize_with = "deserialize_score")]
    pub conflict: Score,
    #[serde(deserialize_with = "deserialize_score")]
    pub dependency: Score,
    pub recent_triggers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FocusRelationshipView {
    #[serde(flatten)]
    pub base: BaseObject,
    pub relationship_edge_id: StableId,
    pub task_type: String,
    pub included_state_refs: Vec<StableId>,
    pub excluded_state_refs: Vec<StableId>,
    pub activation_reason: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InteractionPattern {
    #[serde(flatten)]
    pub base: BaseObject,
    pub participant_role_pattern: String,
    pub trigger: String,
    pub typical_response: String,
    pub variation_rules: Vec<String>,
    pub risk_notes: Vec<String>,
}

fn deserialize_score<'de, D>(deserializer: D) -> Result<Score, D::Error>
where
    D: Deserializer<'de>,
{
    let value = Number::deserialize(deserializer)?;
    let score = value
        .as_f64()
        .ok_or_else(|| D::Error::custom("score must be a finite JSON number"))?;
    if (0.0..=10.0).contains(&score) {
        Ok(value)
    } else {
        Err(D::Error::custom("score must be between 0 and 10"))
    }
}
